File: Algorithms/src/3741. Minimum Distance Between Three Equal Elements II/Code/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var minimumDistance = function (nums) {
  const positions = new Map();

  // Store all indices for each value
  for (let i = 0; i < nums.length; i++) {
    if (!positions.has(nums[i])) {
      positions.set(nums[i], []);
    }
    positions.get(nums[i]).push(i);
  }

  let ans = Infinity;

  // Check every value's index list
  for (const idx of positions.values()) {
    if (idx.length < 3) continue;

    // Check every consecutive group of 3 indices
    for (let i = 0; i + 2 < idx.length; i++) {
      const distance = 2 * (idx[i + 2] - idx[i]);
      ans = Math.min(ans, distance);
    }
  }

  return ans === Infinity ? -1 : ans;
};

File: Algorithms/src/3741. Minimum Distance Between Three Equal Elements II/Code/solution.java
class Solution {
    public int minimumDistance(int[] nums) {
        Map<Integer, List<Integer>> positions = new HashMap<>();
        
        // Store all indices for each value
        for (int i = 0; i < nums.length; i++) {
            positions.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }
        
        int ans = Integer.MAX_VALUE;
        
        // Check every value's index list
        for (List<Integer> idx : positions.values()) {
            if (idx.size() < 3) continue;
            
            // Check every consecutive group of 3 indices
            for (int i = 0; i + 2 < idx.size(); i++) {
                int distance = 2 * (idx.get(i + 2) - idx.get(i));
                ans = Math.min(ans, distance);
            }
        }
        
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }
}